### ISSUE TITLE
Compiler allows non-functions to be bound to active pattern names

### ISSUE THREAD

**Issue by brianrourkeboll** [dotnet/fsharp#17190]
There are a couple of ways in which it is possible to define "active patterns" that are (from the F# perspective) values, not functions. While the tooling treats them as active patterns, it is impossible to actually use them as patterns in any pattern-matching position.

## Expected behavior

The compiler disallows defining a multi-case active pattern that is not a function:

```fsharp
let (|P|Q|) = Choice<unit, unit>.Choice1Of2 ()
// error FS1209: Active pattern '|P|Q|' is not a function
```

I would expect the same for single-case and partial active patterns.

### `(|P|)`

A value bound to what looks like a single-case active pattern name is actually [compiled as a property with a getter](https://sharplab.io/#v2:DYLgZgzgPsCmAuACAFFAClAlIgvIgzEA) (when a module-bound value):

```fsharp
let (|P|) = 3
// val (|P|) : int = 3
```

![image](https://github.com/dotnet/fsharp/assets/14795984/8ecb9af3-da2a-404b-82d8-8bd05b07b6d3)

### `(|P|_|)`

An option or value option value bound to a partial active pattern name is [compiled to the exact same .NET method as it would be if it _were_ defined as a function](https://sharplab.io/#v2:DYLgZgzgPgtg9gEwK7AKYAICCEBqBDYJDAXgFgAodK9NAF3QAooAFKAfSgEp1j0A5OADtUFaulHUA2gB4ATqlpJZgkOgDKtWUgDGtAHwBdCVTqMoARXZce6fIVQDhFCvGRosEAGJJBugJZCPMY0CmasHNwM3LyOIpTUwTLyisqqGlq6hsGmTJYRjNG2BESxQA===), but from the F# perspective it is a value, not a function, although the tooling again treats it as an active pattern.

```fsharp
let (|P|_|) = None
// val (|P|_|) : 'a option
```

![image](https://github.com/dotnet/fsharp/assets/14795984/5b530925-7718-4bbc-9a7a-cd0c99db0cfb)

```fsharp
let (|P|_|) = ValueNone
// val (|P|_|) : 'a voption
```

![image](https://github.com/dotnet/fsharp/assets/14795984/8322973d-4387-4efe-a4bd-c3350fdd375d)

Unfortunately, updating the compiler to disallow these scenarios would be a breaking change — someone could be passing values so defined around _qua_ values (`f (|P|)`, `let r = (|P|) + (|Q|)`, etc.) — although I doubt anyone actually intentionally binds values to active pattern names like this in normal code.

**Comment by edgarfgp** [comment]
A warning behind a lang feature like we have done in other cases might work here ?

**Comment by brianrourkeboll** [comment]
> A warning behind a lang feature like we have done in other cases might work here ?

Yeah, I think that would probably the best we could do.

**Comment by brianrourkeboll** [comment]
It turns out that members with active pattern names are also allowed ([SharpLab](https://sharplab.io/#v2:DYLgZgzgPgLgngBwKYAIBiB7DKC8BYAKBWJQhgEMYBLAYxQFsl6AjJAJxQAooAFKASlwoAzIRKkK1OoxbsuUAIpQA+gKGiiJMpVoMmrDtwBKKtThQA5DADskYrZN0yD8gMqnB5gGrlgAVyQrW3tibSk9WUMoABUoAFUzESA=)) and are _not_ treated as patterns at all by the tooling:

```fsharp
type Foo =
    static member (|P|) = 3
    static member (|Q|_|) = 3
    static member (|R|_|) = None
    static member (|S|_|) = ValueNone
    static member (|T|U|) = 3
```

![image](https://github.com/dotnet/fsharp/assets/14795984/02cd4f45-f0c9-4ba9-8a20-74e372cf102b)

Anonymous records, too:

```fsharp
let x = {| (|P|) = 3 |}
```

And lambdas:

```fsharp
let f : int -> int = fun (|P|) -> (|P|) + (|P|)
```

**Comment by abelbraaksma** [comment]
Yes. from F#'s standpoint, these are just `longident`, and while they have special meaning in some scenarios, they are not illegal. Likewise, if you do have an active pattern, you can just use it as a HOF function argument, or in a function call.

I vaguely remember bringing this up ages ago, but I cannot remember the resolution at the time. We could (should?) issue a warning perhaps, instead of an error?

**Comment by edgarfgp** [comment]
I think we should prevent users from using active pattern are not functions.

**Comment by edgarfgp** [comment]
@vzarytovskii Would be ok to create a new language preview flag e.g. `ImprovedActivePatternReporting` and fix in multiple PR's some of the cases showed in this issues ?

**Comment by vzarytovskii** [comment]
> @vzarytovskii Would be ok to create a new language preview flag e.g. `ImprovedActivePatternReporting` and fix in multiple PR's some of the cases showed in this issue ? 

I suppose, but I will still need to re-read the whole thread to freshen my memory

**Comment by edgarfgp** [comment]
@vzarytovskii  I think we can to start with 

```fsharp
let (|P|) = 3 // This compiles now. But should show `FS1209: Active pattern '|P|' is not a function` 
```
### Why:  
`let (|P|) = 3 `  is not a function and and can not even be used in match

<img width="534" alt="Screenshot 2024-08-13 at 13 30 37" src="https://github.com/user-attachments/assets/4979156e-617c-4fd8-a613-343e0f1523b5">

**Comment by T-Gro** [comment]
<details>
<summary>🤖 Copilot implementation hints</summary>

## Fix

One-line addition in `src/Compiler/Checking/PostInferenceChecks.fs` at `CheckBinding` (~line 2093).

`doesActivePatternHaveFreeTypars` already raises FS1209 for multi-case APs when the type is not a function. The same check is missing for single-case `(|P|)` and partial `(|P|_|)` APs.

In the `match TryGetActivePatternInfo vref with` block, add:
```fsharp
| Some _ ->
    if not (isFunTy g v.TauType) then
        errorR(Error(FSComp.SR.activePatternIdentIsNotFunctionTyped(v.LogicalName), v.Range))
```

**Important:** Use `v.TauType`, not `v.Type` — using `v.Type` produces false positives on generic partial APs like `let (|EqualTo|_|) y x = x = y`.

This is technically a breaking change — gate behind `langversion:preview` or next if required.

## Instructions
- Use TDD: write tests asserting `let (|P|) = 3` and `let (|P|_|) = None` produce FS1209, while `let (|P|_|) x = false` still compiles.
- Run: `dotnet test tests/FSharp.Compiler.ComponentTests/FSharp.Compiler.ComponentTests.fsproj --filter "ActivePattern" -c Debug`
- After code change, invoke the `expert-review` skill to review your work.
- Minimal surgical fix. No bloated comments.
- Run `dotnet fantomas .` on changed files.

</details>
